### ISSUE TITLE
3DS: Tilt/gyro desyncs and resets neutral if moved too fast or too far

### ISSUE THREAD

**Issue by ULTIMAZER0** [TASEmulators/BizHawk#4015]
[//]: # "This description supports Markdown syntax. There's a cheatsheet here: https://guides.github.com/features/mastering-markdown/"
[//]: # "These lines are comments, for letting you know what you should be writing. You can delete them or leave them in."
[//]: # "Also, please don't waste your time writing until you've checked for similar Issues. Remember to check closed Issues too!"

### Summary
[//]: # "Briefly describe what's broken. Include relevant details: loaded core, loaded rom's hash, open tools, running scripts... You can embed a screenshot if it's easier to show the bug, but if you need more than one please put them at the end."
Tilt controls desync and reset neutral if you move too far or too fast. this causes issues in games that need you to tilt to the far side like in kirby triple deluxe...or if you move too quick in OOT3D it causes a random corner to become neutral...causing any time to tilt after that to use the new location as neutral or not move at all.

basically tilt is not working as intended. i did some input recording and on playback it causes desync since its still using true neutral. so if you move too fast it defaults back to 161, 119 so when you record the session you are working around the neutral change in real time....but when it plays back it doesnt have that issue so its not tilting as you recorded it....it playback is the actual true tilt

this.....wouldnt be an issue in TAS recordings or TASstudios where you set the values.....but i do longplays and use the input recording in realtime so when i play it back its one full playthrough

i guess the main issue we run into it....make the realtime gyro controls match what is actually recorded or vice versa.

tl;dr if you tilt too fast (like holding the tilt X to one side and then quickly switching tot he other side or quick circles/eratic movement) it causes the tilt to glitch out and reset neutral to a new location on the screen causing the movie playback to desync because it plays back different than how you recorded it...so we need to make gyro/tilt more stable

### Repro
[//]: # "If you can't figure out the list of steps, delete this section and put 'heisenbug' in the summary somewhere. If a Lua script can cause the bug, you can embed that instead (as simple as possible please)."
1. load any game that uses gyro/tilt controls
2. when in said game....activate the gyro/tilt controls (OOT3D has a really fast instance of being able to use gyro/tilt. right after you gain control of link you can use L1 to enter free look...takes like 2 min while fast forwarding to get to that point)
3. use the tilt button while holding L1 (i made tilt L1 for ease so it does both in one button press) and then use the tilt (if you didnt add tilt set it in the Config > Controllers... > Analog Controls and set Tilt X and Tilt Y) and then just go HAM BANANAS on the tilt x/y and then let the stick go back to neutral. notice that neutral is now idk...in the lower right corner or upper left. then when you go to tilt more from there its using that new neutral location as its center. if you go HAM BANANAS again from there it will choose a new neutral. when you try to compensate for the new neutral the input playback is now offsync because the playback doesnt reset neutral and works as if your NOT compensating for new neutral.
4. for reference....im using L1 to trigger tilt and using the right analogue stick on my controller. i tried mouse for the X and Y too but it also had the same issues

### Host env.
[//]: # "List the computers you've found the bug with. If there's a version that doesn't have the bug, please put that in too. Here are some examples:"
- 2.9.2 dev at https://github.com/TASEmulators/BizHawk/commit/0a3ab1a0555b39605168fb78e68dd670cfad3a0d

NOTE: again...this might not affect or break TASstudios....but it 100% breaks when i input record in realtime for longplays. btw....your program has made longplays so nice. being able to play through save/load states saves me on editing time and play time....thank you all so much for making this. i recently input recorded Zelda Link Between Worlds 3DS longplay and it played back exactly as i recorded it for 12 hours and 40 min of my 14 hour recording before it desync'd. i only had to replay the battle tower and the final dungeon....but then you have Shantae and the Pirates Curse that desyncs every like...30 min of gameplay....haha

thank you all again and i hope to be able to do more 3DS input recording soon (currently working on the 2 Kirbo games but Kirbo Triple Deluxe has ALOT of gyro/tilt sections and the playback ONLY desyncs on the gyro/tilt sections. i have went into TASstudios and put in manual numbers and it works...but that is a TON of work to fix every single gyro/tilt....so im hoping we can get a fix for it soon

**Comment by ULTIMAZER0** [comment]
going HAM BANANAS means flicking the tilt from left to right to left to right....just very quick and sharp movements

**Comment by ULTIMAZER0** [comment]
i have also tried all kinds of different sensitivity/deadzone settings...but they all cause the same issue. sensitivity causes it to not tilt to full and deadzone just expands the neutral...but it needs full tilt range for the games so those settinmgs dont work too well and still cause the desync. just need to make the gyro/tilt more accurate

**Comment by CasualPokePlayer** [comment]
The tilt emulation (as done by BizHawk) is in principle deterministic. On principle too, if it's desyncing in regular movie playback, it would also desync in TAStudio. All of the tilt state too is stored in [savestates](https://github.com/TASEmulators/BizHawk/blob/516a021add659514250056d8b7b995f1114fba0f/src/BizHawk.Emulation.Cores/Consoles/Nintendo/3DS/Encore.IStatable.cs#L30-L46). If anything this is likely just indicating some generic non-determinism if anything (maybe something related to the game's own internal state?)

**Comment by Pixiuchu** [comment]
Could it be that - https://discord.com/channels/280806848909541376/280808167993245707/1424220822477734000 - a message from @CasualPokePlayer in the TASVideos server that links to https://github.com/CasualPokePlayer/encore/blob/cdc97a24baaeb3627009d19467e85aa8a7a51457/src/core/hle/service/hid/hid.cpp#L283 - that this is related?

I'm currently working on a TAS which will require gyroscope movement about 2-3 hours into the TAS, and will require a lot of gyroscope movement at some point after that, so getting this fixed would be pretty nice. I know from prior experience back in 2.10 it desynced a lot (but I wasn't opting for speed so I was able to get a hacky sync that was slow).
